### ISSUE TITLE
Distributed queries is giving inconsistency output between replicas and shards

### ISSUE THREAD

**Issue by Ragavendra-Vigneshwaran-R** [ClickHouse/ClickHouse#57382]
I have 3 shards and 2 replicas of clickhouse cluster and also zookeeper with 3 znodes for replication in my eks cluster. In that, I created a database called "app" with the replicated engine in all shards and replicas (i.e., in all 6 pods). After I created a mergeTree table called "alerts_storage"( which is created in other shards and replicas because of the replicated database engine) for storing the data. On top of it, I have created a distributed table called "alerts" for inserting and retrieving the data evenly in all pods.

Clickhouse config:
`

```
apiVersion: "clickhouse.altinity.com/v1"
kind: "ClickHouseInstallation"
metadata:
  name: "clickhouse"
spec:
  defaults:
    replicasUseFQDN: "no"
    distributedDDL:
      profile: admin_profile
    templates:
      podTemplate: clickhouse-pod
      dataVolumeClaimTemplate: clickhouse-storage
      serviceTemplate: clickhouse-svc
  configuration:
    users:
      default/networks/host_regexp: clickhouse.svc.cluster.local$
      default/networks/ip:
        - 127.0.0.1
        - 0.0.0.0
      admin/networks/host_regexp: clickhouse.svc.cluster.local$
      admin/networks/ip:
        - 127.0.0.1
        - 0.0.0.0
      admin/k8s_secret_password: clickhouse/clickhouse-operator-altinity-clickhouse-operator/password
      admin/access_management: 1
      admin/named_collection_control: 1
      admin/show_named_collections: 1
      admin/show_named_collections_secrets: 1
      admin/profile: admin_profile
      admin/quote: admin_quote
    profiles:
      admin_profile/max_memory_usage: 600000000000
      admin_profile/readonly: 0
      admin_profile/max_insert_threads: 32
    quotas:
      admin_quota/interval/duration: 3600
      admin_quota/interval/queries: 0
      admin_quota/interval/errors: 0
      admin_quota/interval/result_rows: 0
      admin_quota/interval/read_rows: 0
      admin_quota/interval/execution_time: 0
    settings:
      disable_internal_dns_cache: 1
    files:
      conf.d/chop-generated-storage.xml: |
        <yandex>
            <storage_configuration>
                <disks>
                    <default>
                        <metadata_path>/var/lib/clickhouse/</metadata_path>
                    </default>
                    <fast_ssd>
                        <path>/mnt/fast_ssd/clickhouse/</path>
                    </fast_ssd>
                    <s3>
                        <type>s3</type>
                        <endpoint>S3 URL</endpoint>
                        <use_environment_credentials>true</use_environment_credentials>
                        <metadata_path>/var/lib/clickhouse/disks/s3/</metadata_path>
                    </s3>
                </disks>
                <policies>
                    <moving_from_hot_to_cold>
                        <volumes>
                            <default>
                                <disk>default</disk>
                            </default>
                            <hot>
                                <disk>fast_ssd</disk>
                            </hot>
                            <cold>
                                <disk>s3</disk>
                            </cold>
                        </volumes>
                    </moving_from_hot_to_cold>
            	</policies>
            </storage_configuration>
        </yandex>
    zookeeper:
      nodes:
        - host: zookeeper-0.zookeepers.zoo-keeper
          port: 2181
        - host: zookeeper-1.zookeepers.zoo-keeper
          port: 2181
        - host: zookeeper-2.zookeepers.zoo-keeper
          port: 2181
    clusters:
      - name: prod-cluster
        templates:
          podTemplate: clickhouse-pod
          dataVolumeClaimTemplate: clickhouse-storage
        layout:
          shardsCount: 3
          replicasCount: 2
          shards:
            - name: shard0
              internalReplication: "true"
              templates:
                podTemplate: clickhouse-pod
                dataVolumeClaimTemplate: clickhouse-storage
            - name: shard1
              internalReplication: "true"
              templates:
                podTemplate: clickhouse-pod
                dataVolumeClaimTemplate: clickhouse-storage
            - name: shard2
              internalReplication: "true"
              templates:
                podTemplate: clickhouse-pod
                dataVolumeClaimTemplate: clickhouse-storage
  templates:
    serviceTemplates:
      - name: clickhouse-svc
        spec:
          ports:
            - name: http
              port: 8123
            - name: tcp
              port: 9000
          externalTrafficPolicy: Cluster
          type: LoadBalancer
    podTemplates:
      - name: clickhouse-pod
        spec:
          containers:
            - name: clickhouse-pod
              image: clickhouse/clickhouse-server:23.7
              resources:
                limits:
                  cpu: "3"
                  memory: 8Gi
                requests:
                  cpu: "2"
                  memory: 6Gi
              volumeMounts:
                - name: clickhouse-storage
                  mountPath: /var/lib/clickhouse
    volumeClaimTemplates:
      - name: clickhouse-storage
        spec:
          storageClassName: default
          accessModes:
            - ReadWriteOnce
          resources:
            requests:
              storage: 100Gi
```

`
              
 I have inserted million of data in "alerts" table in one month. But when I try select number of count() in alerts table where source='high', it giving inconsistency output in replicas and shards
 
 In 3 pods, chi-clickhouse-prod-cluster-shard0-0-0, chi-clickhouse-prod-cluster-shard1-0-0, chi-clickhouse-prod-cluster-shard2-0-0 it giving below output
 
```
 SELECT count()
FROM alerts
WHERE source = 'high'

Query id: 81a6ed42-5583-4b9d-957b-9a69bccbac40

┌─count()─┐
│    1389 │
└─────────┘

1 row in set. Elapsed: 0.232 sec. Processed 56.50 million rows, 1.02 GB (243.85 million rows/s., 4.39 GB/s.)
```

In 3 pods, chi-clickhouse-prod-cluster-shard0-1-0, chi-clickhouse-prod-cluster-shard1-1-0, chi-clickhouse-prod-cluster-shard2-1-0, it giving below output

```
SELECT count()
FROM alerts
WHERE source = 'high'

Query id: 4c07a903-7ad6-4c92-bbc2-1ac86dcf8691

┌─count()─┐
│    1489 │
└─────────┘

1 row in set. Elapsed: 0.216 sec. Processed 56.96 million rows, 1.03 GB (263.13 million rows/s., 4.74 GB/s.)
```

Why there is inconsistency in output between shards and replicas? Can anyone from clickhouse help here.. what might be cause of this issue?

**Comment by den-crane** [comment]
>mergeTree table called "alerts_storage"

MergeTree or ReplicatedMergeTree?
Provide `create table alerts_storage` `create table alerts`

**Comment by Ragavendra-Vigneshwaran-R** [comment]
@den-crane MergeTree Table. Identified the cause. I checked my alerts data by date wise and found out that this inconsistency in output occurs at 05/11/2023. Before 05/11/2023, the distributed data is giving correct output in both shard and replica. At 05/11/2023, I have restarted zookeeper pod to reduce its memory and cpu limit. That were problem occurs. After restarting the zookeeper, this inconsistency is starts to happen. Is there any way solve this one?

**Comment by lampjian** [comment]
Please reset your "replicasCount: 2" to "replicasCount: 1" if you use MergeTree table engine and query them from a distributed table. Multiple replicas only suitable for Replicated* tables querying from distributed tables consistently.

**Comment by Ragavendra-Vigneshwaran-R** [comment]
@lampjian If I reset the replicasCount to 1, what will happen to the data in 2nd replicas? Will the data will be moved to shards or deleted?

**Comment by lampjian** [comment]
The data in 2nd replica still store somewhere you can not query anymore. It seems to be deleted. You shall transfer the data to 1st replica before you remove the other replica.

**Comment by Ragavendra-Vigneshwaran-R** [comment]
@lampjian  Okay. I have one more doubt. When during resetting the replicaCount to 1 for 2nd replicas, the shards(chi-clickhouse-prod-cluster-shard0-0-0, chi-clickhouse-prod-cluster-shard1-0-0, chi-clickhouse-prod-cluster-shard2-0-0) and it's pvc storage won't be affected, right? They will remains as it is, right? Or will the pvc storage of above shards will be recreated?

**Comment by Ragavendra-Vigneshwaran-R** [comment]
@lampjian Tested in dev environment from the sctrach. Found out that this is not zookeeper problem. From my understanding, it's how the clickhouse actually works with my current settings(shardsCount - 3, replicasCount - 2). 
        
In dev enironment, I created a database called "app" with the replicated engine in all shards and replicas. After I created a mergeTree table called "num_storage"( which is created in other shards and replicas because of the replicated database engine) for storing the data. On top of it, I have created a distributed table called "num" for inserting and retrieving the data evenly.
        
When I inserted 1 to 1000 numbers into the pod chi-clickhouse-prod-cluster-shard0-0-0, the data are distributed evenly inserted across the three shards (chi-clickhouse-prod-cluster-shard0-0-0, chi-clickhouse-prod-cluster-shard1-0-0, chi-clickhouse-prod-cluster-shard2-0-0) only and in its 2nd replicas which is came as independent pod, the data are not inserted.


```
chi-clickhouse-prod-cluster-shard0-0-0.chi-clickhouse-prod-cluster-shard0-0.clickhouse.svc.cluster.local :) select count() from app.num_storage;

SELECT count()
FROM app.num_storage

Query id: 997bd316-c646-49d3-8454-6fea1f7d6e5d

┌─count()─┐
│     351 │
└─────────┘

1 row in set. Elapsed: 0.002 sec.

chi-clickhouse-prod-cluster-shard1-0-0.chi-clickhouse-prod-cluster-shard1-0.clickhouse.svc.cluster.local :) select count() from app.num_storage;

SELECT count()
FROM app.num_storage

Query id: c00fc7d2-359d-462e-9251-4b953eb0698a

┌─count()─┐
│     315 │
└─────────┘

1 row in set. Elapsed: 0.002 sec.

chi-clickhouse-prod-cluster-shard2-0-0.chi-clickhouse-prod-cluster-shard2-0.clickhouse.svc.cluster.local :) select count() from app.num_storage;

SELECT count()
FROM app.num_storage

Query id: 2fbdbe2a-3728-4d7b-a95c-34f2d3c3c8d7

┌─count()─┐
│     334 │
└─────────┘

1 row in set. Elapsed: 0.001 sec.
```
 

When I inserted 1 to 10 numbers into pod chi-clickhouse-prod-cluster-shard0-1-0(which is a 2nd replicas of shard0-0), it data is evenly inserted across the 2nd replicas of shards(chi-clickhouse-prod-cluster-shard0-1-0, chi-clickhouse-prod-cluster-shard1-1-0, chi-clickhouse-prod-cluster-shard2-1-0).
        
 ```
chi-clickhouse-prod-cluster-shard0-1-0.chi-clickhouse-prod-cluster-shard0-1.clickhouse.svc.cluster.local :) select count() from app.num_storage;

SELECT count()
FROM app.num_storage

Query id: 07eb979c-6348-4b6b-8870-c6faa60630e2

┌─count()─┐
│       4 │
└─────────┘

1 row in set. Elapsed: 0.001 sec.

chi-clickhouse-prod-cluster-shard1-1-0.chi-clickhouse-prod-cluster-shard1-1.clickhouse.svc.cluster.local :) select count() from app.num_storage;

SELECT count()
FROM app.num_storage

Query id: 2022a66e-6188-4443-b36c-be685658250c

┌─count()─┐
│       5 │
└─────────┘

1 row in set. Elapsed: 0.001 sec.

chi-clickhouse-prod-cluster-shard2-1-0.chi-clickhouse-prod-cluster-shard2-1.clickhouse.svc.cluster.local :) select count() from app.num_storage;

SELECT count()
FROM app.num_storage

Query id: 00f53697-ae60-4e14-8c21-7f1809db002c

┌─count()─┐
│       1 │
└─────────┘

1 row in set. Elapsed: 0.001 sec.

```


So In conclusion, When we have clickhouse with 3 shards and 2 replicas count and it's use MergeTree table engine and query them from a distributed table on a replicated database, the clickhouse will consider the shards(chi-clickhouse-prod-cluster-shard0-0-0, chi-clickhouse-prod-cluster-shard1-0-0, chi-clickhouse-prod-cluster-shard2-0-0) as one cluster and its 2nd replicas which came as independent pod(chi-clickhouse-prod-cluster-shard0-1-0, chi-clickhouse-prod-cluster-shard1-1-0, chi-clickhouse-prod-cluster-shard2-1-0) as a another cluster. 


Is my conclusion is right? If yes, Could you please tell me what this behaviour of clickhouse is called?

**Comment by den-crane** [comment]
@Ragavendra-Vigneshwaran-R 

>database called "app" with the replicated engine 

replicated database does not replicate data. It synchronize the structure of the table.
You need to use replicated **tables**.

**Comment by Ragavendra-Vigneshwaran-R** [comment]
Okay. I have reseted my replicasCount to 1 after migrating the data from 2nd replica to 1st replicas. Thanks @lampjian @den-crane for you clarification!!! It's really helped for doing the changes in my producation environment without any data loses. Thanks for your help.
